### PR TITLE
[search] Removed one aggressive house number check and replaced another.

### PR DESCRIPTION
**Pull request by mpimenov**
https://jira.mail.ru/browse/MAPSME-1471

This CL slows down the worst case searches.
Below are the results of several runs on queries.txt.

Before:
Maximum response time: 6.546s
Average response time: 0.493s (std. dev. 1.035s)
Maximum response time: 6.713s
Average response time: 0.506s (std. dev. 1.061s)
Maximum response time: 6.548s
Average response time: 0.494s (std. dev. 1.034s)

After:
Maximum response time: 7.735s
Average response time: 0.573s (std. dev. 1.222s)
Maximum response time: 6.528s
Average response time: 0.497s (std. dev. 1.032s)
Maximum response time: 6.877s
Average response time: 0.514s (std. dev. 1.084s)

### CODE REVIEW

**Comment by mpimenov** [comment]
@ygorshenin PTAL

File: search/geocoder.cpp [hunk 1]
```diff
@@ -1095,7 +1095,7 @@ void Geocoder::GreedilyMatchStreets()
       if (IsStreetSynonymPrefix(token))
         continue;
 
-      if (feature::IsHouseNumber(token))
+      if (house_numbers::LooksLikeHouseNumber(token, true /* isPrefix */))
```
Comment (ygorshenin) [inline search/geocoder.cpp]: Please, set up isPrefix correctly here. I.e. you need to check that curToken is the last token and last token is prfix.

**Comment by ygorshenin** [comment]
```R
x1 <- 0.493
sd1 <- 1.035
n1 <- 41

x2 <- 0.573
sd2 <- 1.222
n2 <- 41

t <- (x1 - x2) / sqrt(sd1 ^ 2 / n1 + sd2 ^ 2 / n2)

v1 <- n1 - 1
v2 <- n2 - 1
v <- (sd1 ^ 2 / n1 + sd2 ^ 2 / n2) ^ 2 / (sd1 ^ 4 / (n1 ^ 2 * v1) + sd2 ^ 4 / (n2 ^ 2 * v2))

pvalue <- 2 * pt(-abs(t), df=v)
alpha <- 0.05

if (pvalue < 1 - alpha) {
    message("Test passed")
} else {
    message("Test failed")
}
```

**Comment by ygorshenin** [comment]
It prints "Test passed", therefore we passed Welch t-test, and there're no statistical difference.

**Comment by mpimenov** [comment]
I was more worried about the peak times.

**Comment by ygorshenin** [comment]
Peak times may be quite flaky, query and environment specific. Therefore,
they are not informative. If we need statistics about peak times we need to
conduct a lot of experiments to collect list of peak times, and then do
Welch test on collected data.
16 июня 2016 г. 1:40 пользователь "mpimenov" <notifications@github.com>
написал:

File: search/geocoder.cpp (hunk 1 above)
Comment (mpimenov) [inline search/geocoder.cpp]: done
A test that catches this case would be much appreciated if you have it.

File: search/geocoder.cpp [hunk 2]
```diff
@@ -1095,7 +1095,8 @@ void Geocoder::GreedilyMatchStreets()
       if (IsStreetSynonymPrefix(token))
         continue;
 
-      if (house_numbers::LooksLikeHouseNumber(token, true /* isPrefix */))
+      bool const isPrefix = curToken + 1 == m_numTokens && curToken > m_params.m_tokens.size();
```
Comment (ygorshenin) [inline search/geocoder.cpp]:
```
curToken >= m_params.m_tokens.size()
```

Comment (mpimenov) [inline search/geocoder.cpp]: Right. Sorry for that.

Comment (ygorshenin) [inline search/geocoder.cpp]: It's quite strange that your test didn't catch it.

**Comment by ygorshenin** [comment]
LGTM

**Comment by ygorshenin** [comment]
Thanks for noticing that!

**Pull request merged**